### PR TITLE
Adding RoleArn option for the driver

### PR DESCRIPTION
**Pull request by sarthakkothari**
**Is this a bug fix or adding new feature?**
Adding a new feature

**What is this PR about? / Why do we need it?**
Adding a new startup option that accepts an IAM role that will be used to interact with the EC2 API. The pods need to have permission to assume that role to fetch the credentials.

### CODE REVIEW

File: pkg/cloud/cloud.go [hunk 1]
```diff
@@ -94,6 +96,8 @@ const (
 	volumeDetachedState = "detached"
 	volumeAttachedState = "attached"
 	cacheForgetDelay    = 1 * time.Hour
+
+	assumeRoleDuration = 1 * time.Hour
```
Comment (chu-yik) [inline pkg/cloud/cloud.go]: what is the significance of this duration? what happens after one hour?

File: pkg/cloud/cloud_test.go [hunk 2]
```diff
@@ -110,7 +109,7 @@ func TestNewCloud(t *testing.T) {
 		},
 	}
 	for _, tc := range testCases {
-		ec2Cloud, err := NewCloud(tc.region, tc.awsSdkDebugLog, tc.userAgentExtra, tc.batchingEnabled)
+		ec2Cloud, err := NewCloud(tc.region, tc.awsSdkDebugLog, tc.userAgentExtra, tc.batchingEnabled, "")
```
Comment (chu-yik) [inline pkg/cloud/cloud_test.go:112]: can we also add test case where roleArn is not empty

File: pkg/driver/options.go [hunk 3]
```diff
@@ -32,6 +32,9 @@ type Options struct {
 	// If empty, the in-cluster config will be loaded.
 	Kubeconfig string
 
+	//RoleArn is the role driver will use to interact with the AWS EC2 APIs.
+	RoleArn string
```
Comment (chu-yik) [inline pkg/driver/options.go]: if we are adding this i think we should also update the documentation, specifically https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/docs/options.md

File: pkg/driver/options_test.go [hunk 4]
```diff
@@ -107,6 +110,9 @@ func TestAddFlags(t *testing.T) {
 	if o.ReservedVolumeAttachments != 5 {
 		t.Errorf("unexpected ReservedVolumeAttachments: got %d, want 5", o.ReservedVolumeAttachments)
 	}
+	if o.RoleArn != "arn:aws:iam::012345678910:role/ExampleRole" {
+		t.Errorf("unexpected role-arn: got %d, want arn:aws:iam::012345678910:role/ExampleRole", o.ReservedVolumeAttachments)
```
Comment (chu-yik) [inline pkg/driver/options_test.go]: `o.ReservedVolumeAttachments` does not seem correct, did you mean `o.RoleArn`

File: pkg/driver/options.go (hunk 3 above)
Comment (chu-yik) [reply pkg/driver/options.go]: Also i don't see this being set anywhere - don't we have to set this somewhere in https://github.com/sarthakkothari/aws-ebs-csi-driver/blob/e296e801d93b86ed811f28069813e1cb8c88224d/pkg/driver/options.go#L95

File: pkg/cloud/cloud.go (hunk 1 above)
Comment (sarthakkothari) [reply pkg/cloud/cloud.go]: The credentials sts vends is valid for an hour. After an hour it refreshes. Default is 15 mins. This is what was done for Karpenter so I used it, happy to remove it.

File: pkg/cloud/cloud_test.go (hunk 2 above)
Comment (sarthakkothari) [reply pkg/cloud/cloud_test.go:112]: I'm not sure what it will test? It just is a struct that gets attached to `aws.Config.Credentials` which is then used to create a new EC2 service client (https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/b8e3563db111522638620374a100146e2272b293/pkg/cloud/cloud.go#L358). This test is just tests are just testing conversions from internal package representation of `cloud` object to `Cloud` interface which is exposed for other packages in the directory.

File: pkg/driver/options.go [hunk 5]
```diff
@@ -91,6 +94,7 @@ type Options struct {
 
 func (o *Options) AddFlags(f *flag.FlagSet) {
 	f.StringVar(&o.Kubeconfig, "kubeconfig", "", "Absolute path to a kubeconfig file. The default is the emtpy string, which causes the in-cluster config to be used")
+	f.StringVar(&o.RoleArn, "role-arn", "", "Arn of the role to be used while interacting with EC2 APIs. The default is the emtpy string, which causes the role provided by the Pod identity or OIDC to be used.")
```
Comment (chu-yik) [inline pkg/driver/options.go]: nit: typo `emtpy` -> `empty`

File: pkg/cloud/cloud_test.go (hunk 2 above)
Comment (chu-yik) [reply pkg/cloud/cloud_test.go:112]: we need a test for the logic added in `cloud.go` L351-L355 to catch regression - without this we risk in a future release where the logic is broken but not caught.

maybe we can use a mock STS client and assert that assume role is invoked with the expected role ARN? it doesn't necessary have to be in the cloud_test.go unit test, but we need SOME test for this logic, maybe explore if we can add a test under `e2e/`, generally what we want to assert is - when a roleArn is provided, the credential from that particular role is used.

File: pkg/cloud/cloud.go (hunk 1 above)
Comment (chu-yik) [reply pkg/cloud/cloud.go]: Thanks for explaining! - i see we are referencing ENO from https://code.amazon.com/packages/EKSNodeOperator/blobs/9702a7dd7e23e968170f46a3eabe44b92128635a/--/pkg/awsclients/provider.go#L14,L47,L48,L49, consider naming this as `sessionDuration`, that might make more sense.


just to confirm - with this implementation, after the session expires in an hour, the assume role provider would be able to assume the role again?

File: pkg/cloud/cloud.go [hunk 6]
```diff
@@ -333,17 +337,23 @@ var _ Cloud = &cloud{}
...
+func newEC2Cloud(region string, awsSdkDebugLog bool, userAgentExtra string, batchingEnabled bool, roleArn string) Cloud {
 	cfg, err := config.LoadDefaultConfig(context.Background(), config.WithRegion(region))
 	if err != nil {
 		panic(err)
 	}
 
+	if roleArn != "" {
+		cfg.Credentials = stscreds.NewAssumeRoleProvider(sts.NewFromConfig(cfg), roleArn, func(aro *stscreds.AssumeRoleOptions) {
+			aro.Duration = assumeRoleDuration
+		})
```
Comment (chu-yik) [inline pkg/cloud/cloud.go]: not too sure about directly setting `cfg.Credentials` here, so `cfg` was created using the default config - which corresponds to a role (A) in our service account, this role would allow us to assume the customer provided role (B)

would changing the `cfg` impact how the AssumeRoleProvider tries to assume role again after the session expires in an hour? i.e. would the STS client now also attempt to use credential for role (B) (that has expired) to assume the role (B) again?

Comment (chu-yik) [reply pkg/cloud/cloud.go]: would it make more sense to do sth similar to ENO here and create an `aws.Config` for customer role? i'm thinking something like this

create ec2Cfg using the role here https://github.com/sarthakkothari/aws-ebs-csi-driver/blob/e6a95cbd67fb7e1612c4a0a4839a526b273cf054/pkg/cloud/cloud.go#L351
```
if roleArn != "" {
  ec2Cfg := // create the aws.Config
} else {
  ec2Cfg := cfg
}
```

and use this when creating ec2API in https://github.com/sarthakkothari/aws-ebs-csi-driver/blob/e6a95cbd67fb7e1612c4a0a4839a526b273cf054/pkg/cloud/cloud.go#L368

```
svc := ec2.NewFromConfig(ec2Cfg, // ...
```

File: pkg/cloud/cloud.go (hunk 1 above)
Comment (sarthakkothari) [reply pkg/cloud/cloud.go]: Yes, that is what I've been informed by my team

File: pkg/cloud/cloud.go (hunk 6 above)
Comment (sarthakkothari) [reply pkg/cloud/cloud.go]: I understand what you're saying, my thoughts would be what if there are other things in the `cfg` that are used by the EC2 which we are not aware of?

File: pkg/cloud/cloud_test.go (hunk 2 above)
Comment (sarthakkothari) [reply pkg/cloud/cloud_test.go:112]: Yeah, I will need some feedback from connor on this because all of this is just setting the providers, the actual calls are not being made for me to mock.
